tests/v2/validation/public_api/crdgeneration/crd_generation.go

```diff
@@ -0,0 +1,138 @@
+	fleetlocal           = "fleet-local"
+	fleetdefault         = "fleet-default"
+	crdJsonFilePath      = "../resources/crds.json"
+	roleTemplateJsonPath = "../resources/roleTemplate.json"
+)
+
+func mapCRD(crdList []string) map[string][]string {
+	groups := []string{groupManagement, groupCatalog, groupUI, groupCluster, groupProject, groupRKE, groupProvisioning, groupFleet}
+	crdMap := make(map[string][]string)
+
+	for _, group := range groups {
+		crdMap[group] = []string{}
+	}
+
+	crdMap["additional"] = []string{}
+
+	for _, crd := range crdList {
+		found := false
+		for _, group := range groups {
+			if strings.Contains(crd, group) {
+				crdMap[group] = append(crdMap[group], crd)
+				found = true
+				break
+			}
+		}
+		if !found {
+			crdMap["additional"] = append(crdMap["additional"], crd)
+		}
+	}
+
+	return crdMap
+}
+
+func listCRDS(client *rancher.Client, clusterID string) ([]string, error) {
+	crdCollection, err := customresourcedefinitions.ListCustomResourceDefinitions(client, clusterID, "")
+	if err != nil {
+		return nil, err
+	}
+
+	crds := make([]string, len(crdCollection.Items))
+	for idx, crd := range crdCollection.Items {
+		crds[idx] = crd.GetName()
+	}
+	return crds, nil
+}
+
+func validateCRDList(t *testing.T, crdsList []string, crdMapPreUpgrade map[string][]string, clusterName string) {
+	crdMap := mapCRD(crdsList)
+
+	if clusterName == "local" {
+		for group, value := range crdMap {
+			switch group {
+			case groupManagement:
+				assert.Equal(t, crdMapPreUpgrade[groupManagement], value)
+				assert.Equal(t, len(crdMapPreUpgrade[groupManagement]), len(value))
+			case groupCatalog:
+				assert.Equal(t, crdMapPreUpgrade[groupCatalog], value)
+				assert.Equal(t, len(crdMapPreUpgrade[groupCatalog]), len(value))
+			case groupUI:
+				assert.Equal(t, crdMapPreUpgrade[groupUI], value)
+				assert.Equal(t, len(crdMapPreUpgrade[groupUI]), len(value))
+			case groupProject:
+				assert.Equal(t, crdMapPreUpgrade[groupProject], value)
+				assert.Equal(t, len(crdMapPreUpgrade[groupProject]), len(value))
+			case groupRKE:
+				assert.Equal(t, crdMapPreUpgrade[groupRKE], value)
+				assert.Equal(t, len(crdMapPreUpgrade[groupRKE]), len(value))
+			case groupProvisioning:
+				assert.Equal(t, crdMapPreUpgrade[groupProvisioning], value)
+				assert.Equal(t, len(crdMapPreUpgrade[groupProvisioning]), len(value))
+			case groupFleet:
+				assert.Equal(t, crdMapPreUpgrade[groupFleet], value)
+				assert.Equal(t, len(crdMapPreUpgrade[groupFleet]), len(value))
+			}
+		}
+	}
+}
+
+func validateCRDDescription(t *testing.T, client *rancher.Client, clusterV1 *v1.Cluster, clusterID string) {
+
+	logs, err := kubectl.KubectlExplain(client, clusterV1, roleTemplate, clusterID)
+	require.NoError(t, err)
+	assert.NotEmpty(t, logs)
+	desIndexStart := strings.Index(logs, "DESCRIPTION:")
+	desIndexEnd := strings.Index(logs, "FIELDS")
+	if desIndexStart == -1 {
+		t.Fatalf("Pod logs do not contain description field")
```

Comment:
would this be a drastic error? We should not use fatals, this would affect code coverage as well unless this is a breaking error. Maybe a forced require error so the test ends here.

Reply:
I wouldn't say its a drastic error here. I can change it to require.Error(), but could you let me know why/how it would affect the code coverage ?

Reply:
Fatals are for extreme cases that will exit the program immediately, and not safety.  Doing so will affect our code coverage process, because we need a graceful exit to be able to write test reports. Especially in testing we should refrain from using fatals at all.